metals/src/main/scala/scala/meta/internal/metals/PackageProvider.scala

```diff
@@ -443,24 +548,123 @@ class PackageProvider(
+      tree: Tree,
+      oldPkgParts: List[String],
+  ): List[TopLevelDeclaration] = {
+    def go(
+        tree: Tree,
+        currentPkgs: List[String] = Nil,
+    ): LazyList[TopLevelDeclaration] = tree match {
+      case d: Defn with Member =>
+        val isCaseClass = d match {
+          case c: Defn.Class =>
+            c.mods.collectFirst { case _: Mod.Case => () }.isDefined
+          case _ => false
+        }
+        val additionalPositions = d match {
+          case c: Defn.Class if c.mods.collectFirst { case _: Mod.Implicit =>
+                ()
+              }.isDefined =>
+            d.children.flatMap(go(_, currentPkgs)).map(_.pos)
+          case _ => List.empty
+        }
+        LazyList(
+          TopLevelDeclaration(
+            d.name.value,
+            currentPkgs,
+            d.name.pos.toLsp,
+            isCaseClass,
+            additionalPositions,
+          )
+        )
+      case Defn.Val(_, pats, _, _) =>
+        LazyList.from(pats).collect { case Pat.Var(name) =>
+          TopLevelDeclaration(name.value, currentPkgs, name.pos.toLsp)
+        }
+      case Defn.Var(_, pats, _, _) =>
+        LazyList.from(pats).collect { case Pat.Var(name) =>
+          TopLevelDeclaration(name.value, currentPkgs, name.pos.toLsp)
+        }
+      case g: Defn.GivenAlias =>
+        LazyList(
+          TopLevelDeclaration(g.name.value, currentPkgs, g.name.pos.toLsp)
+        )
+      case p: Pkg =>
+        LazyList
+          .from(p.children)
+          .flatMap(go(_, currentPkgs ++ extractNames(p.ref)))
+      case p: Pkg.Object =>
+        LazyList
+          .from(p.children)
+          .flatMap(go(_, currentPkgs ++ List(p.name.value)))
+      case t => LazyList.from(t.children).flatMap(go(_, currentPkgs))
     }
+    val oldPkgPartsLength = oldPkgParts.length
     go(tree).toList
+      .map(decl => decl.copy(name = decl.name.dropPkgParts(oldPkgPartsLength)))
+  }
+
+  private def extendPositionToIncludeNewLine(pos: inputs.Position) = {
+    if (pos.start > 0 && pos.input.chars(pos.start - 1) == '\n')
+      inputs.Position.Range(pos.input, pos.start - 1, pos.end)
+    else pos
   }
 
   private def findRefs(
-      name: Name,
+      decl: TopLevelDeclaration,
       uri: String,
-  ): List[(String, Location)] = {
-    val params = new ReferenceParams(
+  ): List[Reference] = {
+    def params(pos: Range) = new ReferenceParams(
       new TextDocumentIdentifier(uri),
-      new Position(name.pos.startLine, name.pos.startColumn),
+      pos.getStart(),
       new ReferenceContext(false), // do not include declaration
     )
-    refProvider.references(params).flatMap(s => s.locations.map((s.symbol, _)))
+    (decl.pos :: decl.additionalPosForReferenceSearch)
+      .flatMap(pos => refProvider.references(params(pos)))
+      .flatMap(s =>
+        s.locations.map { loc =>
+          val sym = s.symbol
+          val symbols =
+            if (decl.isCaseClass) List(s"${sym.dropRight(1)}.", sym)
```

Comment:
Why is `sym.dropRight(1)` needed?

Reply:
So this is a dumb hack. For case class the symbols is `Class#`, but really we get an object too, so we artificially produce `Class.`, `sym.dropRight(1) ++ "."`. We use those symbols later to see if a wildcard import imports something from a different file as well, so any reference to this synthetic object would stop us from deleting this wildcard import.